packages/server/modules/workspaces/services/workspaceJoinRequests.ts

```diff
@@ -38,23 +47,38 @@ export const requestToJoinWorkspaceFactory =
     createWorkspaceJoinRequest,
     sendWorkspaceJoinRequestReceivedEmail,
     getUserById,
-    getWorkspace
+    getWorkspaceWithDomains,
+    getUserEmails
   }: {
     createWorkspaceJoinRequest: CreateWorkspaceJoinRequest
     sendWorkspaceJoinRequestReceivedEmail: SendWorkspaceJoinRequestReceivedEmail
     getUserById: GetUser
-    getWorkspace: GetWorkspace
+    getWorkspaceWithDomains: GetWorkspaceWithDomains
+    getUserEmails: FindEmailsByUserId
   }) =>
   async ({ userId, workspaceId }: { userId: string; workspaceId: string }) => {
     const requester = await getUserById(userId)
     if (!requester) {
       throw new NotFoundError('User not found')
     }
 
-    const workspace = await getWorkspace({ workspaceId })
+    const workspace = await getWorkspaceWithDomains({ id: workspaceId })
     if (!workspace) {
       throw new WorkspaceNotFoundError('Workspace not found')
     }
+    if (!workspace?.discoverabilityEnabled) throw new WorkspaceNotDiscoverableError()
+    const workspaceDomains = workspace.domains.filter((domain) => domain.verified)
+    if (!workspaceDomains.length) throw new WorkspaceNotJoinableError()
+
+    const userEmails = await getUserEmails({ userId })
+    const matchingEmail = userEmails.find((userEmail) => {
+      if (!userEmail.verified) return false
+      return workspaceDomains
+        .map((domain) => domain.domain)
+        .includes(userEmail.email.split('@')[1])
+    })
+
+    if (!matchingEmail) throw new WorkspaceJoinNotAllowedError()
```

Comment:
I think, there might be another error for the domain policy mismatch

Reply:
this all logic was taken from https://github.com/specklesystems/speckle-server/blob/main/packages/server/modules/workspaces/services/join.ts#L15 

Which will be removed once we deploy this new flow